Your name here.  
Your section number here.  

# Workshop 1: Canopy, python basics, and a little plotting

** Submit this notebook to bCourses to receive a grade for this Workshop.**

Please complete workshop activities in code cells in this iPython notebook. Include comments in your code when necessary. Enter your name in the cell at the top of the notebook, and rename the notebook [email_name]_Workshop01.ipynb, where [email_name] is the part of your UCB email address that precedes "@berkeley.edu" 

## Installing Canopy

You'll want to have python installed on your computer for this course, so here are some instructions to do that. While the download and installation are running, feel free to continue on with the rest of the workshop. You can run python code in your browser [here (coding ground)](http://www.tutorialspoint.com/ipython_terminal_online.php) or [here (trinket)](https://trinket.io/). Hopefully at least one of those sites works for you!

### Getting the download started

Go to the Enthought Canopy website to [download the installer](https://store.enthought.com/downloads/#default) here. Check that the operating system and 32/64 bit options at the top are correct before hitting download. The default options are probably good. If you're not sure whether your computer is 32-bit or 64-bit (most modern computers are 64-bit), here are directions to find out for [Windows](https://esupport.trendmicro.com/en-us/home/pages/technical-support/1038680.aspx), [Mac](http://www.chiefarchitect.com/support/article/KB-01230/determining-if-your-computer-is-32-bit-or-64-bit.html), and [Linux](http://www.howtogeek.com/198615/how-to-check-if-your-linux-system-is-32-bit-or-64-bit/).



### Making an account

While that's downloading, request an academic license to access all the features of Canopy. Go to [this page](https://store.enthought.com/#canopy-academic) and click "Request your license". Fill in your information to make an account. Once you've installed Canopy, you'll be able to "Log in" to access Canopy's suite "Training on Demand" tutorials along with various other features within the program.

### Installation

Once the installer finishes downloading, installation should be straightforward. (This is one of the reasons we chose Canopy, after all.) Simply open the downloaded file and follow the default options. Once it's installed, open Canopy and once again follow the default options to set up your python environment. If you'd like more specific install instructions, Canopy has OS-specific guides here:
* [Windows Installation](http://docs.enthought.com/canopy/quick-start/install_windows.html)
* [Mac OSX Installation](http://docs.enthought.com/canopy/quick-start/install_macos.html)
* [Linux Installation](http://docs.enthought.com/canopy/quick-start/install_linux.html)

### Updating packages

That should be about it! You can make sure everything is up to date by opening the "Package Manager" from the main Canopy window. Click "Updates" (in the left sidebar) and then "Install All Updates" in the bottom right. If there are several updates, it might take a while to finish. Leave that working in the background and check back on it later.

## The Terminal / Command Line / Shell

You will find it useful to know some basics about the command line (aka terminal or shell) on your computer. The command line lets you navigate around your computer and perform tasks without any graphical user interface---just text. You don't need a mouse to use the command line, just your keyboard. With experience, some people find it faster to use command line tools than graphical tools to do certain things on their computer.

So let's dive in! If you're on a Mac or Linux, find and open the Terminal. If you're on Windows, go to your start menu and type "cmd" (no quotes), then run it.

You'll be faced with a prompt where you can type things in. Press `Enter` to execute whatever you've typed. Try typing these commands (in this order if you'd like!) to get a feel for your terminal. Type `ls` or `dir` occasionally as you change things to see how the contents of your folder have changed.

See where you are
    - On Mac/Linux:  pwd
    - On Windows:    cd

See what's in the current directory
    - On Mac/Linux:  ls
    - On Windows:    dir
    
Change directories (into the `Documents` directory, for instance)
    - On Mac/Linux:  cd Documents
    - On Windows:    cd Documents
    
Go back to the "parent directory"
    - On Mac/Linux:  cd ..
    - On Windows:    cd ..
    
Make a directory (the directory `fooDirectory`, for instance)
    - On Mac/Linux:  mkdir fooDirectory
    - On Windows:    mkdir fooDirectory

Make another directory!
    -                mkdir barDirectory
    
Move something (move `barDirectory` into `fooDirectory` directory, for instance)
    - On Mac/Linux:  mv barDirectory fooDirectory
    - On Windows:    move barDirectory fooDirectory

Check the contents of a different directory
    - On Mac/Linux:  ls fooDirectory
    - On Windows:    dir fooDirectory
    
Change through multiple directories at once
    - On Mac/Linux:  cd fooDirectory/barDirectory
    - On Windows:    cd fooDirectory\barDirectory
        (notice different slash directions!)

Go back to your "home" directory
    - On Mac/Linux:  cd
                or:  cd ~
    - On Windows:    cd %userprofile%
    
Rename something (rename `fooDirectory` into `RenamedDir`, for instance)
    - On Mac/Linux:  mv fooDirectory RenamedDir
    - On Windows:    move fooDirectory RenamedDir

Remove an empty directory (the directory `myDirectory`, for instance)
    - On Mac/Linux:  rmdir fooDirectory/barDirectory
    - On Windows:    rmdir fooDirectory\barDirectory
    
Learn more about a command (the `cd` command, for instance)
    - On Mac/Linux:  man cd
    - On Windows:    cd /?

Print the contents of a file to the terminal (`file.txt`, for instance)
    - On Mac/Linux:  cat file.txt
    - On Windows:    type file.txt

## Using the Canopy editor

### IPython Interpreter

Time to write your first python code! Open the Canopy "Editor". You should see a couple buttons to create or open files, and maybe a IPython interpreter at the bottom. (If you don't see anything at the bottom, click "View" -> "Python" and it should appear.) Type in "2+2" (without the quotes ... or with them!) and see what happens! Next, try typing these (you don't have to type any of the "# comments!") ...

    x = "Phys"
    y = "ics!"
    z = x + y     # Put 'em together
    z             # See what we got!
    y + x         # Backwards!
    len(z)        # 8 characters in total ...
    len(z)**2     # Computing the area?
    z[0]          # Grab the first character
    z[1:3]        # Grab the next two characters
    z[:4]
    z[:4] == x    # Test a match!
    z[4:] == y
    z[:]          # The whole string
    z[::-1]       # The whole string, right to left
    z[1::3]       # Start at the second character and take every third character from there
    z*3 + 5*z[-1] # Woo!

This interpreter is a great place to test things out and mess around. In Canopy, it also comes preloaded with a couple libraries (like numpy and matplotlib) that we will use a lot in this course. Try typing `log(e)`, for instance, and you should get `1.0`. Both the function `log` and the number `e` are from the `numpy` library. Or type `pie([1,2,3])`, since `pie` is defined by matplotlib! This can be convenient, but also confusing since many names and variables are already used in ways you might not expect. When writing scripts you'll have to manually import any library you want to use. This little inconvenience is greatly worth the confusion it can save.

### Writing your first script

Let's write a script now! Type the following code snippet into the code cell below:
    
    import numpy as np
    
    'Hello world! Have some pie!', np.pi
    'And some pie from a log!', np.log(np.pi)

In [3]:
import numpy as np

'Hello world! Have some pie!', np.pi
'And some pie from a log', np.log(np.pi)

('And some pie from a log', 1.1447298858494002)

This should produce ... nothing. Why not? In the interactive shell, when you type a string or a number, like `'Physics!'` or `np.pi`, it will immediately print the result. (try it!) In a script, if you want python to display something for you, you have to tell it. For strings, we use the `print` function. (Later for plots, you'll learn to use matplotlib's `show()` function.) Adjust your code to read:

    import numpy as np
    
    print 'Hello world! Have some pie!', np.pi
    print 'And some pie from a log!', np.log(np.pi)

This should produce

    Hello world! Have some pie! 3.14159265359
    And some pie from a log! 1.14472988585

In [4]:
import numpy as np

print 'Hello world! Have some pie!', np.pi
print 'And some pie from a log!', np.log(np.pi)

Hello world! Have some pie! 3.14159265359
And some pie from a log! 1.14472988585


Much better! Using what you learned in lecture, try changing this program to print only 4 decimal places in each number. The result should look like:

    Hello world! Have some pie! 3.1416
    And some pie from a log! 1.1447

In [5]:
import numpy as np

print 'Hello world! Have some pie',round(np.pi,4)
print 'and some pie from a log', round(np.log(np.pi), 4)


Hello world! Have some pie 3.1416
and some pie from a log 1.1447


## Debugging Practice

[Adapted from Langtangen, Exercise 1.16] Working with a partner, type these statements into your python interpreter. Figure out why some statements fail and correct the errors.

*Hint: Try testing the left- and right-hand sides seperately before you put them together in statements. It's ok if you don't understand yet why some of the expressions evaluate to the results they do (like the last one).*

    1a = 2
    a1 = b
    x = 2
    y = X + 4    # is it 6?
    5 = 5    # is it True?
    4/5 == 4.0/5.0   # is it True?
    10/2 == 10/2.   # is it True?
    type(10/2) == type(10/2.)   # is it True?

In [29]:
#1a = 2 invalid syntax
a=2 
#a1 = b
a1 = 'b'
x = 2
#y = X + 4    # is it 6?
y = x+4
#5 = 5    # is it True? cant assing literal to literal
5==5
#4/5 == 4.0/5.0   # is it True? No, false
4/5 == int(4.0/5.0) 
#10/2 == 10/2.   # is it True? No false
10/2 == int(10/2.)
type(10/2) == type(10/2.)   # is it True?No false


False

    from Math import factorial
    print factorial(pi)
    pi = "3.14159'
    c = 4**3**2**3
    _ = ((c-78564)/c + 32))
    discount = 12%
    AMOUNT = 120.-
    amount = 120$

In [26]:
from Math import factorial
print factorial(pi)
pi = '3.14159'
c = 4**3**2**3
_ = ((c-78564)/(c + 32))
discount = .12
AMOUNT = "20
amount = 120

ImportError: No module named Math

    address = hpl@simula.no
    and = duck
    class = 'INF1100, gr 2'
    continue_ = x > 0
    Erdoğan = """Erdoğan"""
    rev = fox = True
    Norwegian = ['a human language']
    true = fox is rev in Norwegian

In [28]:
address = "hpl@simula.no"
and = "duck"
class = 'INF1100, gr 2'
continue_ = x > 0
Erdoğan = """Erdoğan"""
rev = fox = True
Norwegian = ['a human language']
true = fox is rev in Norwegian

SyntaxError: invalid syntax (<ipython-input-28-a3d65456b644>, line 2)

## Four Fours

[Inspired by Harvey Mudd College's CS5 course] Here's an arithmetic game to try your hand at. Your task is to compute each of the numbers, from 1 through 11, using exactly four 4's and simple math operations. You're allowed to use `+` (addition), `-` (subtraction), `*` (multiplication), `/` (division), `sqrt()` (square root), `factorial()` (factorial), and `%` (modulus). You're also allowed to use `.4` (that's one 4) or `44` (that's two 4's) if you'd like. Just remember, you must use exactly four 4 digits total!

As a reminder, four factorial (denoted by $!$ in mathematics) is $4! = 4 \cdot 3 \cdot 2 \cdot 1$, and the modulus operator (usually denoted by $\text{mod}$ in mathematics) is the remainder after division. For instance, $\ 5\ \text{mod}\ 2 = 1$, $\ 13\ \text{mod}\ 7 = 6$, and $\ 14\ \text{mod}\ 7 = 0$.

Copy and paste the following code to start:

    from math import factorial, sqrt
    
    print 'Zero:', 4 - 4 + 4 - 4
    print 'One:'
    print 'Two:'
    print 'Three:'
    print 'Four:'
    print 'Five:'
    print 'Six:'
    print 'Seven:'
    print 'Eight:'
    print 'Nine:'
    print 'Ten:'
    print 'Eleven:'

We've given you `zero` for free, as `4 - 4 + 4 - 4`. Of course, we could have also done `44 * (.4 - .4)` or `factorial(4) - 4 * (4 + sqrt(4))`, since both of those also yield `0` (or rather, `0.0`. Why is that?) and use exactly four 4's.

In [20]:
from math import factorial, sqrt

print 'Zero:', 4 - 4 + 4 - 4
print 'One:', (4*4)/(4*4)
print 'Two:', (4*4)/(4+4)
print 'Three:', sqrt(4*4)-(4/4) #this returns 3.0 because of the return type of the sqrt function
print 'Four:',4% ( 4*4*4)
print 'Five:', (4*4+4) / 4
print 'Six:' , ((4+4)/4) + 4
print 'Seven:',(4 + 4 - 4/4)
print 'Eight:', 4%(4*4)+4
print 'Nine:', (4+4/4) + 4
print 'Ten:', (4*4)-sqrt(4)-4 #same reason as #3
print 'Eleven:', 44 / (sqrt(4)*sqrt(4)) # sqrt 4

Zero: 0
One: 1
Two: 2
Three: 3.0
Four: 4
Five: 5
Six: 6
Seven: 7
Eight: 8
Nine: 9
Ten: 10.0
Eleven: 11.0


Your final source code will be full of four fours formulas, but your final output should look like this:

    Zero: 0
    One: 1
    Two: 2
    Three: 3
    Four: 4
    Five: 5
    Six: 6
    Seven: 7
    Eight: 8
    Nine: 9
    Ten: 10
    Eleven: 11

It's ok if some of these have a trailing `.0` (`0.0`, for instance), but make sure you understand why they do!

## Canopy Tutorials and Exercises

Let's take a look at the Canopy tutorials and exercises, which are excellent resources for you as you learn and practice python in this course and beyond. In the main Canopy window, click "Training on Demand". Log in to Canopy using the button in the upper-right of the page, if you aren't logged in already. The page you should then see is a gateway into all of Canopy's training materials, which are freely accessible to you as an academic.

Click on "Python Essentials". Here you see video lectures for many topics, some which you've already learned, some which we'll get to later, and some that we won't have time to cover in this course. Check these out sometime if you'd ever like a reminder or a fresh perspective on the material.

For now, click on the "Exercises" tab above the video thumbnails, and click on "star data". It should open an IPython notebook for you in Canopy automatically, but if it doesn't, you can click "(download zip)" to download the files and then open them in canopy yourself.

Do the "star data" exercise for some practice extracting data from files. 

## Plotting Practice: Envelope and Oscillation

[Adapted from Ayars, Problem 0-0] Paste the following code into a code cell to start:

    import numpy as np
    import matplotlib.pyplot as plt
    
    plt.figure()    # This makes sure you start with a clean figure each time your code is run
    
    ### Your code here!
    
    plt.show()      # This tells python to display the plot you've made
    
Finish the program to graph both of the following functions on a single figure, with a usefully-sized scale.

1. $ x^4 e^{-2x} $
1. $ \left[ x^2 e^{-x} \sin(x^2) \right]^2 $

It may help to refer to the lecture notes or Ayars Chapter 0. Use numpy's `linspace` to make an array for the $x$ values, and use numpy's `exp` and `sin` functions to make arrays for the $y$ values of the two expressions. Make sure the $x$ values are spaced close enough together to make the curves smooth! You can label the axes as simply $x$ and $y$, and add whatever title you'd like. Add some appropriate legend. For instance, `plt.legend(['envelope', 'oscillation'])` would be fine for this workshop.

In [17]:
import numpy as np
import math
import matplotlib.pyplot as plt

plt.figure()    # This makes sure you start with a clean figure each time your code is run
#plt.plot([1,2,3], [3,5,6])



### Your code here!
def graph(formula , x_range):
    x = np.array(x_range)
    y=eval(formula)
    plt.plot(x,y)
graph ("x^2")
#graph("(x**4*e**(-2x))", range(0, 10))
#graph("(x**2*e**(-x)sin(x**2))**2", range(0,10))


plt.show()      # This tells python to display the plot you've made


#### Hint
In the editor, type something like `np.linspace(` or `np.linspace()` and then hit the TAB key. The editor should pop up with documentation for that function, to remind you of what it does and how to use it. This works for all sorts of functions if you ever forget how to use them! Google is also your friend for all manner of questions about python, matplotlib, numpy, ...

#### Just for fun
For a pretty mathematical legend, you can add this line instead:

    plt.legend([r'$x^4 e^{-2x}$', r'$[x^2 e^{-x} \sin(x^2)]^2$'])

Writing a string as `r'$<LaTeX code here>$'` tells matplotlib to process the strings as LaTeX math code. LaTeX is a programming language for typesetting pretty documents. You'll want to learn it at some point in your physics career, but you don't need to worry about it now.

## Plotting Practice: Radioactivity Data

[Adapted from Ayars, Problem 0-2] Paste the following code into a code cell to start:

    import numpy as np
    import matplotlib.pyplot as plt
    
    ### Load the data here
    
    plt.figure()    # Start a clean figure for your first plot
    
    ### Your code for the first plot here!
    
    plt.figure()    # Start a clean figure for your second plot
    
    ### Your code for the second plot here!
    
    plt.show()      # This tells python to display the plots you've made
    

The file Ba137.txt contains two columns. The first is counts from a Geiger counter, the second is time in seconds. (available here: https://raw.githubusercontent.com/celegante/code_chapter_0-_github/master/Ba137.txt . Right-click the link to save it as a .txt file.) 

1. Make a useful graph of this data, with axes labels and a title.
2. If this data follows an exponential curve, then plotting the natural log of the data (or plotting the raw data on a logarithmic scale) will result in a straight line. Determine whether this is the case, and explain your conclusion with---you guessed it---an appropriate graph.

Be sure to add comments throughout your code so it's clear what each section of the code is doing! It may help to refer to the lecture notes or Ayars Chapter 0.

Try using `'x'` or `'^'` as the marker type in your `plt.plot()` functions (instead of `'g-'`, for instance), to get a single x or triangle for each data point instead of a connected line. Google if you'd like to learn more options!

Once you're through, your code should produce two graphs, one with the data, another with the natural log of the data, both labelled appropriately. It should also print out a clear answer to the question in part 2 (e.g., `Yes, the data follows an exponential curve`, or `No, the data does not follow an exponential curve`).

In [19]:
import numpy as np
import matplotlib.pyplot as plt

### Load the data here
plt.figure()
x, y = np.loadtxt('ba137.txt', unpack = True)
plt.plot(x, y, label = "Data")

plt.xlabel("Time (s)")
plt.ylabel("Number of decays (N)")
plt.title("BA 137 half life")

axes = plt.gca()
axes.set_xlim([0,40000])
axes.set_ylim([0,600])
plt.legend()


plt.show()      # This tells python to display the plots you've made


#### Hints

Put the file in the same directory as your python file, and use numpy's `loadtxt` or `genfromtxt` function to load each column into an array for use in your plots. 

If your file isn't loading correctly, it might be because your IPython working directory isn't the same as the directory your script and Ba137.txt are in.)

If you'd like to learn more about what `loadtxt` does (or why the `unpack = True` option is important), type `loadtxt?` or `help(loadtxt)` into the python interpreter for documentation. Press `q` to get out of the documentation.

## You're done!

Congratulations, you've finished this week's workshop! You're welcome to leave early or get started on this week's homework.